core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostService.scala

```diff
@@ -17,34 +19,86 @@ object SubmissionCostService {
+    if( workflowIds.isEmpty ) {
+      Future.successful(Map.empty[String, Float])
+    } else {
+      for {
+        //try looking up the workflows via the submission ID.
+        //this makes for a smaller query string (though no faster).
+        submissionCosts <- executeSubmissionCostsQuery(submissionId, workspaceNamespace, submissionDate)
+        //if that doesn't return anything, fall back to
+        fallbackCosts <- if (submissionCosts.size() == 0)
+          executeWorkflowCostsQuery(workflowIds, workspaceNamespace, submissionDate)
+        else
+          Future.successful(submissionCosts)
+      } yield {
+        extractCostResults(fallbackCosts)
+      }
+    }
+  }
 
-    extractWorkflowCostResults(executeWorkflowCostsQuery(workflowIds, workspaceNamespace))
+
+  def getWorkflowCost(workflowId: String,
+                      workspaceNamespace: String,
+                      submissionDate: Option[DateTime]): Future[Map[String, Float]] = {
+    executeWorkflowCostsQuery(Seq(workflowId), workspaceNamespace, submissionDate) map extractCostResults
   }
 
   /*
    * Manipulates and massages a BigQuery result.
    */
-  def extractWorkflowCostResults(rowsFuture: Future[util.List[TableRow]]): Future[Map[String, Float]] = {
-
-    rowsFuture map { rowsOrNull =>
-      Option(rowsOrNull) match {
-        case Some(rows) => rows.asScala.map { row =>
-          // workflow ID is contained in the 2nd cell, cost is contained in the 3rd cell
-          row.getF.get(1).getV.toString -> row.getF.get(2).getV.toString.toFloat
-        }.toMap
-        case None => Map.empty[String, Float]
-      }
+  def extractCostResults(rows: util.List[TableRow]): Map[String, Float] = {
+    Option(rows) match {
+      case Some(rows) => rows.asScala.map { row =>
+        // workflow ID is contained in the 2nd cell, cost is contained in the 3rd cell
+        row.getF.get(1).getV.toString -> row.getF.get(2).getV.toString.toFloat
+      }.toMap
+      case None => Map.empty[String, Float]
+    }
+  }
+
+  private def partitionDateClause(submissionDate: Option[DateTime]): String = {
+    (submissionDate map { d: DateTime =>
+      val date = d.toString(DateTimeFormat.forPattern("yyyy-MM-dd"))
+      s"AND _PARTITIONDATE >= $date"
+    }).getOrElse("")
+  }
+
+  private def executeSubmissionCostsQuery(submissionId: String, workspaceNamespace: String, submissionDate: Option[DateTime]): Future[util.List[TableRow]] = {
+
+    val querySql: String =
+      s"""SELECT wflabels.key, REPLACE(wflabels.value, "cromwell-", "") as `workflowId`, SUM(billing.cost)
+      |FROM `$tableName` as billing, UNNEST(labels) as wflabels
+      |CROSS JOIN UNNEST(billing.labels) as blabels
+      |WHERE blabels.value = "terra-$submissionId"
+      |AND wflabels.key = "cromwell-workflow-id"
+      |AND project.id = ?
+      |${partitionDateClause(submissionDate)}
+      |GROUP BY wflabels.key, workflowId""".stripMargin
```

Comment:
I will doublecheck this query once some test runs have been exported to BQ. It's a bit odd to be doing a cross join but it's what gets both the submission AND workflow labels in separate columns. If there are other more sane approaches I would love to hear them!

Reply:
I checked it and it works.